helix-view/src/view.rs

```diff
@@ -576,6 +614,47 @@ impl View {
             self.apply(&transaction, doc);
         }
     }
+
+    pub fn get_backup_theme(&self, theme: &Theme) -> usize {
+        let selection_scope = theme
+            .find_scope_index("ui.selection")
+            .expect("could not find `ui.selection` scope in the theme!");
+        selection_scope
+    }
+
+    pub fn get_search_matches(
+        &self,
+        theme: &Theme,
+    ) -> Option<Vec<(usize, std::ops::Range<usize>)>> {
+        let selection_scope = match theme.find_scope_index("ui.selection.primary") {
+            Some(theme) => theme,
+            None => self.get_backup_theme(theme),
+        };
+
+        let mut ret = Vec::new();
+        for (start, end) in &self.search_matches.matches {
+            ret.push((selection_scope, *start..*end));
+        }
+
+        if ret.is_empty() {
+            None
+        } else {
+            Some(ret)
+        }
+    }
+
+    pub fn update_search_matches(&mut self, contents: &str, regex: &Regex) {
+        // this is expensive. only update search matches if search criteria has changed
+        if self.search_matches.has_changed(regex) {
+            self.search_matches.update(regex);
+            self.search_matches.clear();
+            for one_match in self.search_matches.new.find_iter(contents) {
+                self.search_matches
+                    .matches
+                    .push((one_match.start(), one_match.end()));
+            }
+        }
+    }
```

Comment:
the first `if` checks if the `Regex` search criteria has changed. this prevents this block from executing if the user is traversing match results using `n` or similar. if the `Regex` search criteria has indeed changed, `old` is set equal to `new`, and existing matches are cleared. new matches are calculated via expensive `Regex` operation based on `new` search criteria